### PR TITLE
SONAR-8192 add root system administer backward compatibility

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java [hunk 1]
```diff
@@ -58,6 +62,9 @@ public boolean apply(DbSession dbSession, UserPermissionChange change) {
       default:
         throw new UnsupportedOperationException("Unsupported permission change: " + change.getOperation());
     }
+    if (SYSTEM_ADMIN.equals(change.getPermission())) {
+      dbClient.userDao().updateRootFlagFromPermissions(dbSession, change.getUserId().getLogin(), defaultOrganizationProvider.get().getUuid());
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java]: I recommend that backend like DAO always works with technical keys but not functional keys. Here I suggest to use user id (`change.getUserId().getId()`).

Side-note: as you commented last week, `UserId` should indeed be renamed `UserRef`. I'd do that soon.

File: sonar-plugin-api/src/main/java/org/sonar/api/utils/internal/AlwaysIncreasingSystem2.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+import org.sonar.api.utils.System2;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * A subclass of {@link System2} which implementation of {@link System2#now()} always return a bigger value than the
+ * previous returned value.
```
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/utils/internal/AlwaysIncreasingSystem2.java]: As this class is in main code, I suggest to add a comment to highlight that it's reserved to tests.

File: sonar-db/src/test/java/org/sonar/db/user/UserDaoTest.java [hunk 3]
```diff
@@ -49,13 +50,16 @@
...
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
+import static org.sonar.core.permission.GlobalPermissions.SYSTEM_ADMIN;
 import static org.sonar.db.user.GroupMembershipQuery.IN;
 import static org.sonar.db.user.GroupMembershipQuery.builder;
 import static org.sonar.db.user.GroupTesting.newGroupDto;
 import static org.sonar.db.user.UserTesting.newUserDto;
 
 public class UserDaoTest {
   private static final long NOW = 1500000000000L;
```
Comment (simonbrandhof) [inline sonar-db/src/test/java/org/sonar/db/user/UserDaoTest.java]: the value could be formatted with thousand separator, as you did for DATE_1 and DATE_2

File: server/sonar-server/src/main/java/org/sonar/server/permission/GroupPermissionChanger.java [hunk 4]
```diff
@@ -74,9 +78,16 @@ private boolean removePermission(DbSession dbSession, GroupPermissionChange chan
       change.getOrganizationUuid(),
       change.getGroupIdOrAnyone().getId(),
       change.getNullableProjectId());
+    updateRootFlag(dbSession, change);
     return true;
   }
 
+  private void updateRootFlag(DbSession dbSession, GroupPermissionChange change) {
+    if (SYSTEM_ADMIN.equals(change.getPermission()) && !change.getGroupIdOrAnyone().isAnyone()) {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/permission/GroupPermissionChanger.java]: the root flag relates to global permissions only. You can ignore modifications of project permissions by adding the condition on `!change.getProjectId().isPresent()`.

File: server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java (hunk 1 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/permission/UserPermissionChanger.java]: the root flag relates to global permissions only. You can ignore modifications of project permissions by adding the condition on `!change.getProjectId().isPresent()`.

File: sonar-db/src/test/java/org/sonar/db/user/UserDaoTest.java (hunk 3 above)
Comment (sns-seb) [inline sonar-db/src/test/java/org/sonar/db/user/UserDaoTest.java]: not my code, but ok

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 1 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by sns-seb** [comment]
> The root flag must be updated when a group is deleted (api/usergroups/delete). If the group had system permission, then some users may loose their root status.

good catch, I'm taking care of it

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

**Pull request merged**